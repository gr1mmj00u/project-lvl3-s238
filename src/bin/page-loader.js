#!/usr/bin/env node
import commander from 'commander';
import pageLoader from '..';

commander
  .version('1.0.2')
  .arguments('<dir> <url>')
  .description('Compares two configuration files and shows a difference.')
  .option('-o, --output', 'output format')
  .action((dir, url, options) => {
    if (typeof dir === 'undefined') {
      console.error('no dir parameter!');
      process.exit(1);
    }
    if (typeof url === 'undefined') {
      console.error('no url parameter!');
      process.exit(1);
    }
    if (options.output) {
      pageLoader(url, dir);
    } else {
      pageLoader(url);
    }
  });

commander.parse(process.argv);
