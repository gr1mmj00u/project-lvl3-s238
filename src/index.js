import pageloader from './page-loader';

export default pageloader;
