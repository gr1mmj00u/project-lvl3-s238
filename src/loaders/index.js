import renderAst from './ast';
import renderPlain from './plain';
import renderJson from './json';

const renders = {
  ast: renderAst,
  plain: renderPlain,
  json: renderJson,
};

export default format => (data) => {
  const render = renders[format];
  if (!render) {
    throw new Error(`unkown render format: ${format}`);
  }
  return render(data);
};
