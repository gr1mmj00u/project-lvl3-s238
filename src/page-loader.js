import axios from 'axios';
import mzfs from 'mz/fs';
import Listr from 'listr';
import url from 'url';
import path from 'path';
import cheerio from 'cheerio';
import debug from 'debug';

const log = debug('page-loader:main');

const selectors = {
  img: 'src',
  script: 'src',
  link: 'href',
};

export const getFileName = (link, ext = '') => {
  const { host, path: pathLink } = url.parse(link);
  const { name, dir, ext: fileExt } = path.parse(pathLink);
  const parseUrl = (host)
    ? `${host}${path.resolve(dir, name)}`
    : `${path.resolve(dir, name)}`;

  const extention = (ext) ? `.${ext}` : fileExt;

  const fileName = parseUrl.match(/(\w*)/g)
    .filter(e => e)
    .join('-');

  return `${fileName}${extention}`;
};

const loadResource = (link, dir, fileName) => {
  log(`Log resourse:
    link = %s;
    dir = %s;
    fileName = %s`, link, dir, fileName);
  const fullFileName = path.resolve(dir, fileName);

  return {
    title: link,
    task: () => axios.get(link, { responseType: 'arraybuffer' })
      .catch((e) => {
        log(e);
        throw new Error(`Resource by (${link}) was not saved.`);
      })
      .then(response => mzfs.writeFile(fullFileName, response.data))
      .then(() => path.parse(fullFileName)),
  };
};

const getTagsObjects = ($, tags) => tags.reduce((acc, tag) => {
  const elements = $(tag).get();
  return [...acc, ...elements];
}, []);

const parsePage = (data, hostLink, assetsFolder, assetsFolderPath) => {
  const $ = cheerio.load(data);
  const tagObjects = getTagsObjects($, ['img', 'script', 'link']);

  const tasks = tagObjects.reduce((acc, e) => {
    const attribute = selectors[e.name];
    const ref = $(e).attr(attribute);
    const { host } = url.parse(ref);

    if (host) {
      return acc;
    }

    const fileName = getFileName(ref);
    const link = url.resolve(hostLink, ref);
    const newLink = path.resolve('/', assetsFolder, fileName);

    log('New link %s', newLink);

    $(e).attr(attribute, newLink);

    return [...acc, loadResource(link, assetsFolderPath, fileName)];
  }, []);

  const loadResourseTasks = new Listr(tasks);
  const promiseListr = Promise.resolve(loadResourseTasks.run());

  return mzfs.mkdir(assetsFolderPath)
    .then(() => promiseListr)
    .then(() => $.html());
};

export default (link, dir) => {
  const { protocol, host } = url.parse(link);
  const rootLink = url.format({ protocol, host });

  const pageName = getFileName(link, 'html');
  const fullPageName = path.resolve(dir, pageName);

  const assetsFolder = `${getFileName(link)}_files`;
  const assetsFolderPath = path.resolve(dir, assetsFolder);

  if (!mzfs.existsSync(dir)) {
    return Promise.reject(new Error(`Folder (${dir}) does not exists.`));
  }

  return axios.get(link, { responseType: 'arraybuffer' })
    .catch((e) => {
      log(e);
      throw new Error(`Resource by (${link}) was not saved.`);
    })
    .then(response => mzfs.writeFile(fullPageName, response.data))
    .then(() => mzfs.readFile(fullPageName, 'utf-8'))
    .then(data => parsePage(data, rootLink, assetsFolder, assetsFolderPath))
    .then(html => mzfs.writeFile(fullPageName, html));
};
