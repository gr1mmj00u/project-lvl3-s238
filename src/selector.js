const selectors = {
  link: 'link[href]',
  img: 'img[src]',
  script: 'script[src]',
};

export default tag => ($) => {
  const selector = selectors[tag];
  if (!selector) {
    throw new Error(`unkown selector: ${tag}`);
  }
  return $(selector).map((i, e) => $(e)).get();
};
