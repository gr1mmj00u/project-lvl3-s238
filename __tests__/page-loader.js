import fs from 'mz/fs';
import os from 'os';
import path from 'path';
import nock from 'nock';
import axios from 'axios';
import debug from 'debug';
import httpAdapter from 'axios/lib/adapters/http';
import pageLoader from '../src';
import { getFileName } from '../src/page-loader';

const getFullPath = file => path.resolve(__dirname, file);
const log = debug('page-loader:test');

const tmpdir = os.tmpdir();
const host = 'http://www.hexlet.io';
const query = '/test/save/page/?test=10&param=20';
const pageLink = `${host}${query}`;
const sourcePage = getFullPath('__fixtures__/source.html');
const expectedPage = getFullPath('__fixtures__/expected.html');

axios.defaults.adapter = httpAdapter;

beforeEach(() => {
  nock(host)
    .get(query)
    .replyWithFile(200, sourcePage, { 'Content-Type': 'text/html' });

  nock(host)
    .get('/src/css/test.css')
    .replyWithFile(
      200,
      path.resolve(__dirname, '__fixtures__/src/css/test.css'),
      { 'Content-Type': 'text/css' },
    );

  nock(host)
    .get('/src/js/test.js')
    .replyWithFile(
      200,
      path.resolve(__dirname, '__fixtures__/src/js/test.js'),
      { 'Content-Type': 'application/javascript' },
    );

  nock(host)
    .get('/src/img/image1.jpg')
    .replyWithFile(
      200,
      path.resolve(__dirname, '__fixtures__/src/img/image1.jpg'),
      { 'Content-Type': 'image/jpeg' },
    );

  nock(host)
    .get('/src/img/image2.png')
    .replyWithFile(
      200,
      path.resolve(__dirname, '__fixtures__/src/img/image2.png'),
      { 'Content-Type': 'image/png' },
    );
  nock(host)
    .get('/src/img/image3.png')
    .reply(404);
});

test('Save page', async () => {
  const expected = await fs.readFile(expectedPage, 'utf-8');
  try {
    const tempPath = await fs.mkdtemp(path.join(tmpdir, 'foo-'));
    await pageLoader(pageLink, tempPath);

    const fullfileName = path.resolve(tempPath, getFileName(pageLink, 'html'));
    const data = await fs.readFile(fullfileName, 'utf-8');

    expect(data).toBe(expected);
  } catch (e) {
    log('Save page => %s', e.toString());
    expect(e).toMatch('error');
  }
});

test('Save files', async () => {
  const css = await fs.readFile(getFullPath('__fixtures__/src/css/test.css'), 'utf-8');
  const js = await fs.readFile(getFullPath('__fixtures__/src/js/test.js'), 'utf-8');
  const jpeg = await fs.readFile(getFullPath('__fixtures__/src/img/image1.jpg'), 'utf-8');
  const png = await fs.readFile(getFullPath('__fixtures__/src/img/image2.png'), 'utf-8');

  try {
    const tempPath = await fs.mkdtemp(path.join(tmpdir, 'foo-'));
    await pageLoader(pageLink, tempPath);

    const assetsPath = path.resolve(tempPath, `${getFileName(pageLink)}_files`);
    const cssData = await fs.readFile(path.resolve(assetsPath, 'src-css-test.css'), 'utf-8');
    const jsData = await fs.readFile(path.resolve(assetsPath, 'src-js-test.js'), 'utf-8');
    const jpegData = await fs.readFile(path.resolve(assetsPath, 'src-img-image1.jpg'), 'utf-8');
    const pngData = await fs.readFile(path.resolve(assetsPath, 'src-img-image2.png'), 'utf-8');

    expect(cssData).toBe(css);
    expect(jsData).toBe(js);
    expect(jpegData).toBe(jpeg);
    expect(pngData).toBe(png);
  } catch (e) {
    log('Save files => %s', e.toString());
    expect(e).toMatch('error');
  }
});

test('404 status', async () => {
  try {
    const tempPath = await fs.mkdtemp(path.join(tmpdir, 'foo-'));
    await pageLoader(path.resolve(host, '/src/img/image3.png'), tempPath);
  } catch (e) {
    log('404 status => %s', e.toString());
    expect(e.message).toMatch('Resource by (/src/img/image3.png) was not saved.');
  }
});

test('Output dir is not exist', async () => {
  try {
    await pageLoader(path.resolve(host, '/src/img/image3.png'), './test/folder/');
  } catch (e) {
    log('Output dir is not exist => %s', e.toString());
    expect(e.message).toMatch('Folder (./test/folder/) does not exists.');
  }
});
