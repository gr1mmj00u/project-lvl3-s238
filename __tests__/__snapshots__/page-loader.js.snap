// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`404 status 1`] = `"Resource by (/src/img/image3.png) was not saved."`;

exports[`Output dir is not exist 1`] = `"Folder (./test/folder/) does not exists."`;
